### PR TITLE
#54: added support for passing object as value

### PR DESCRIPTION
**Pull request by fkrafi**

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -1,20 +1,28 @@
...
+}
+
 export default function useLocalStorage(
   key: string,
-  initialValue: string = ''
-): [string, Dispatch<string>] {
-  const [value, setValue] = useState(
-    () => window.localStorage.getItem(key) || initialValue
+  initialValue: string | object = ''
+): [string | object, Dispatch<string | object>] {
```
Comment (TheAifam5) [inline src/index.ts]: Rather an string or an pure object, what do you thing about adding generic parameter type, like returning T, where T extends from an object?

File: src/index.ts [hunk 2]
```diff
@@ -1,20 +1,28 @@
 import { Dispatch, useCallback, useEffect, useState } from 'react';
 
+function getValueFromLocalStorage(key: string) {
+  const value = window.localStorage.getItem(key);
+  return value ? (value.startsWith('{') ? JSON.parse(value) : value) : null;
```
Comment (TheAifam5) [inline src/index.ts]: Identifying based on first character is not a best approach, user may store some arbitrary data starting with that character, like format string. Whats the fallback scenario when it starts with { but does not end (invalid json object)?

File: src/index.ts [hunk 3]
```diff
@@ -1,20 +1,28 @@
...
+    () => getValueFromLocalStorage(key) || initialValue
   );
 
-  const setItem = (newValue: string) => {
+  const setItem = (newValue: string | object) => {
     setValue(newValue);
-    window.localStorage.setItem(key, newValue);
+    window.localStorage.setItem(
+      key,
+      typeof newValue === 'string' ? newValue : JSON.stringify(newValue)
```
Comment (TheAifam5) [inline src/index.ts]: If the string is created using "new String('value')", the condition will fail. You can add additional check using "instanceof" or use different approach:

Object.prototype.toString.call(newValue) == '[object String]'

Additionally, "null" and "undefined" will also fail I think.

File: test/index.test.ts [hunk 4]
```diff
@@ -46,3 +46,32 @@ describe('useLocalStorage', (): void => {
...
+
+  describe('Setup', () => {
+    it('Returns initial object value', () => {
+      localStorage.removeItem(KEY);
+      const { result } = renderHook(() => useLocalStorage(KEY, VALUE.INITIAL));
+      expect(result.current[0]).toEqual(VALUE.INITIAL);
+    });
+
+    it('When no initial object value is passed, returns an empty string', () => {
+      localStorage.removeItem(KEY);
```
Comment (TheAifam5) [inline test/index.test.ts]: You can move this to "beforeEach" scope.

File: test/index.test.ts [hunk 5]
```diff
@@ -46,3 +46,32 @@ describe('useLocalStorage', (): void => {
     expect(localStorage.getItem(KEY)).toBe(VALUE.CHANGED);
   });
 });
+
+describe('useLocalStorage with object', (): void => {
+  const KEY = 'key';
+  const VALUE = {
+    INITIAL: { key1: 'value1' },
+    CHANGED: { key1: 'value1', key2: 'value2' },
```
Comment (TheAifam5) [inline test/index.test.ts:54]: This is unused / test is missing.

**Comment by TheAifam5** [comment]
In this StackBlitz example, I've made a playground around mutable and immutable objects
https://stackblitz.com/edit/react-ts-gi66ra?file=index.tsx

The first one, if the value is not saved in localStorage, the view does not get updated and only one last character is stored (in this example, initial + the last entered character) in localStorage.

The other ones works well with mutable and immutable objects.

Could you check that one too why does it happen?

Additionally the README should be updated too ;)

**Review by TheAifam5 — CHANGES_REQUESTED** [review]

File: src/index.ts (hunk 2 above)
Comment (fkrafi) [reply src/index.ts:5]: can i use `lodash` package

**Comment by fkrafi** [comment]
@TheAifam5 , i have changed all the reviews from your comments, can you please verify?

**Pull request closed without merging**